verification.md

```diff
@@ -11,24 +12,296 @@ module VERIFICATION
+
+The following lemmas essentially capture the signature extraction mechanisms. It reduces the reasoning efforts of the underlying theorem prover, factoring out the essence of the byte-twiddling operations.
+
+```{.k .uiuck}
+  // for viper
+  rule #padToWidth(N, #asByteStack(#asWord(WS))) => WS
+    requires noOverflow(WS) andBool N ==Int #sizeWordStack(WS)
+
+  // for solidity
+  rule #asWord(WS) /Int D => #asWord(#take(#sizeWordStack(WS) -Int log256Int(D), WS))
+    requires D %Int 256 ==Int 0 andBool D >=Int 0
+     andBool #sizeWordStack(WS) >=Int log256Int(D)
+     andBool noOverflow(WS)
+
+  // syntactic sugar
+  syntax Bool ::= noOverflow(WordStack)    [function]
+                | noOverflowAux(WordStack) [function]
+  rule noOverflow(WS) => #sizeWordStack(WS) <=Int 32 andBool noOverflowAux(WS)
+  rule noOverflowAux(W : WS) => 0 <=Int W andBool W <Int 256 andBool noOverflowAux(WS)
+  rule noOverflowAux(.WordStack) => true
+```
+
+### Abstraction for Hash
+
+We do not model the hash function as an injective function simply because it is not true due to the pigeonhole principle.
+
+Instead, we abstract it as an uninterpreted function that captures the possibility of the hash collision, even if the probability is very small.
+
+```
+  syntax Int ::= keccak ( WordStack ) [function, smtlib(smt_keccak)]
+```
+
+However, one can avoid reasoning about the potential collision by assuming all of the hashed values appearing in each execution trace are collision-free. This can be achieved by instantiating the injectivity property only for the terms appearing in the symbolic execution, in a way analogous to the universal quantifier instantiation.
+
+#### Hashed Locations
+
+Hashed locations are essential for the storage layout schemes used to store compound data structures such as maps in the storage. For more details, refer to [EVM-Specific EVM Specification](proofs/erc20.md#storage).
+
+The following syntactic sugars capture the storage layout schemes of Solidity and Viper.
+
+```{.k .uiuck}
+  syntax IntList ::= List{Int, ""} [klabel(intList)]
+
+  syntax Int ::= #hashedLocation(String, Int, IntList) [function]
+  rule #hashedLocation(LANG, BASE, .IntList) => BASE
+  //
+  rule #hashedLocation("Viper",    BASE, OFFSET OFFSETS) => #hashedLocation("Viper",    keccakIntList(BASE) +Word OFFSET, OFFSETS)
+  rule #hashedLocation("Solidity", BASE, OFFSET OFFSETS) => #hashedLocation("Solidity", keccakIntList(OFFSET BASE),       OFFSETS)
+
+  syntax Int ::= keccakIntList(IntList) [function]
+  rule keccakIntList(VS) => keccak(intList2ByteStack(VS))
+
+  syntax WordStack ::= intList2ByteStack(IntList) [function]
+  rule intList2ByteStack(.IntList) => .WordStack
+  rule intList2ByteStack(V VS) => #asByteStackInWidth(V, 32) ++ intList2ByteStack(VS)
+    requires 0 <=Int V andBool V <Int (2 ^Int 256)
+```
```

Comment:
Don't take absolutely what I said before, that we should do things separately for Viper and Solidity.  Just think about it.  It may be better to keep them together after all ...